### ISSUE TITLE
Run perf on Windows

### ISSUE THREAD

**Issue by rylev** [rust-lang/rustc-perf#834]
It would be useful to also have the ability to run the perf suite on Windows. One open question would be whether this would run as often as the Linux tests for or if we would kick off Windows tests at a different cadence or on-demand. 

This might also help with #181

**Comment by Mark-Simulacrum** [comment]
I don't personally have familiarity with Windows, but a key piece we'd need is to determine what tooling to use in place of `perf` on Linux.

**Comment by wesleywiser** [comment]
I've started working on this. Based on [these](https://randomascii.wordpress.com/2016/11/27/cpu-performance-counters-on-windows/) [blog](https://easyperf.net/blog/2019/02/23/How-to-collect-performance-counters-on-Windows) [posts](https://adamsitnik.com/Hardware-Counters-ETW/), it looks like we can capture a variety of hardware performance counters on Windows. I've also confirmed that this technique works and I can see instructions retired for rustc. 

Right now, I'm trying to get the collector building and running on Windows. I'm a bit curious @Mark-Simulacrum if you have any thoughts about how to organize the platform dependent code.

**Comment by Mark-Simulacrum** [comment]
Hm, so I think there are two big, related, questions here:

* Do we want to collect statistics on perf.rust-lang.org for Windows?
   * Do we want to "double up" statistics, or perhaps only benchmark crates like winapi or windows-rs on Windows?

I'm going to operate on the assumption that the answer to the first is potentially yes, but the answer to the second is no - we don't actually want to be constantly comparing instruction counts on linux vs windows, for example. That means that likely no or very little work needs to be done on the frontend and database schema, and we can limit to the perf-stat invocation supporting Windows equivalents.

I think my recommendation is that we don't try to add another perf collector, but rather substitute the backend in https://github.com/rust-lang/rustc-perf/blob/master/collector/src/rustc-fake.rs#L44 and try to mangle the relevant outputs in that file such that the majority of rustc-perf code doesn't even need to know it's running with "windows perf" vs. "linux perf". Does that sound plausible?

Of course, if there are more bits that are linux-specific I'm not thinking of right now, we should talk about those.

**Comment by rylev** [comment]
> Do we want to "double up" statistics

I'd like to see this land before we get too deep into discussing strategies of when to run the Windows tests. I agree that it's likely overkill to run them as often as we run Linux tests, but I do however want us to run them in some sort of automated way, and I do think it's interesting to run benchmarks that only make sense on Windows like winapi/windows-rs as often as we run the existing benchmarks. 

Of course, this all leads to interesting questions of which hardware we'll use to run these tests.

**Comment by wesleywiser** [comment]
> Do we want to collect statistics on perf.rust-lang.org for Windows?

I think that would be great but we'll need to explore how to make that actually happen of course. 

> Do we want to "double up" statistics, or perhaps only benchmark crates like winapi or windows-rs on Windows?

I think there is some value in "doubling up" some of the crates, not to compare Linux vs Windows but just to make sure we aren't introducing regressions that affect Windows disproportionately for some reason. For now though, I would be very happy just to get windows-rs being benchmarked. 

> That means that likely no or very little work needs to be done on the frontend and database schema, and we can limit to the perf-stat invocation supporting Windows equivalents.

Yeah, I agree. I'm focusing on just being able to invoke the collector on Windows.

> Of course, if there are more bits that are linux-specific I'm not thinking of right now, we should talk about those.

I was more referring to stuff like this:

https://github.com/rust-lang/rustc-perf/blob/c52ee623e231e7690a93be88d943016968c1036b/collector/src/execute.rs#L26-L68

where we shell out to Linux utilities. If it's feasible, I'm planning to convert these to Rust code or otherwise use standard Windows tools. I would rather not require userland Linux tools to be available to run the collector.

**Comment by Mark-Simulacrum** [comment]
Yes, those should all likely be converted to Rust code - if something proves difficult to convert, we can continue shelling out to a platform-specific utility but hopefully that's not necessary for the most part.

I agree that getting support landed for running the collector manually is the first priority, just wanted to communicate that if we wanted to extend the official collection - particularly to double up testing - we'd likely need to do some database/frontend work as well, depending on the strategy used.
